### PR TITLE
Hawkey package build test (fix #76)

### LINKED ISSUE #76: try build hawkey package with RPG
follow the process of RPG (all the phases: see /tmp/rpg-*) and watch warnings -> add issues/fix it
the testing project is libsolv project folder. (Install libsolv rpm to fullfill dependencies)

### PR DESCRIPTION
**Pull request by LukasSlouka**
Local build seems okay, lets see what travis has to say about it

### CODE REVIEW

**Comment by LukasSlouka** [comment]
@cibo94 for some reason `test_plugins` failed, I suspect it is due to 9d5eb02c3e6c77a608eec4ff93523cc7aedc1362

**Comment by LukasSlouka** [comment]
```
...F.....
======================================================================
FAIL: test_find_files (test_plugins.FindPatchPluginTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/Lukas/Documents/Redhat/rpg/tests/unit/test_plugins.py", line 105, in test_find_files
    sorted_files)
AssertionError: Lists differ: [('/M[555 chars], ('/patch/0.patch', None, None), ('/patch/1.p[489 chars]one)] != [('/M[555 chars], ('/mock_project/__pycache__/', '%exclude', N[586 chars]one)]

First differing element 14:
('/patch/0.patch', None, None)
('/mock_project/__pycache__/', '%exclude', None)

Second list contains 2 additional elements.
First extra element 28:
('/translation/CZ.mo', None, None)

Diff is 1287 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
Ran 9 tests in 0.036s

FAILED (failures=1)
```

**Comment by LukasSlouka** [comment]
Doing this issue uncovered some problems.

1. As far as changing `spec.BuildRequires` and other `sets`, I noticed that in for example `test_build_rpg.py` we are doing union the wrong way (`self.base.spec.BuildRequires.union...`) whereas to change it we have to `self.base.spec.BuildRequires = self.base.spec.BuildRequires.union...`. I dont know the extent of this, but project should be checked for it and debugged. 

2 - In my opinion there is some issue with excluding/including `.pyc`, `.pyo` files and `__pycache__` directories. I tried to build rpm of hawkey manually calling
```bash
mock --no-clean --no-cleanup-after --verbose --rebuild hawkey-0.6.0-1.fc22.src.rpm --resultdir ~/rpmbuild/RPMS
```
on rpg-generated SRPM. Build went okay but stopped with this error:

```
Checking for unpackaged file(s): /usr/lib/rpm/check-files /builddir/build/BUILDROOT/hawkey-0.6.0-1.fc22.x86_64
error: Installed (but unpackaged) file(s) found:
   /usr/lib64/python2.7/site-packages/hawkey/__init__.pyc
   /usr/lib64/python2.7/site-packages/hawkey/__init__.pyo
   /usr/lib64/python2.7/site-packages/hawkey/test/__init__.pyc
   /usr/lib64/python2.7/site-packages/hawkey/test/__init__.pyo
    Directory not found: /builddir/build/BUILDROOT/hawkey-0.6.0-1.fc22.x86_64/usr/lib64/python2.7/site-packages/hawkey/__pycache__
    Directory not found: /builddir/build/BUILDROOT/hawkey-0.6.0-1.fc22.x86_64/usr/lib64/python2.7/site-packages/hawkey/test/__pycache__
    Directory not found: /builddir/build/BUILDROOT/hawkey-0.6.0-1.fc22.x86_64/usr/local/include/hawkey/__pycache__
    Directory not found: /builddir/build/BUILDROOT/hawkey-0.6.0-1.fc22.x86_64/usr/local/lib64/__pycache__
    Directory not found: /builddir/build/BUILDROOT/hawkey-0.6.0-1.fc22.x86_64/usr/local/lib64/pkgconfig/__pycache__
    Directory not found: /builddir/build/BUILDROOT/hawkey-0.6.0-1.fc22.x86_64/usr/local/share/man/man3/__pycache__
    Installed (but unpackaged) file(s) found:
   /usr/lib64/python2.7/site-packages/hawkey/__init__.pyc
   /usr/lib64/python2.7/site-packages/hawkey/__init__.pyo
   /usr/lib64/python2.7/site-packages/hawkey/test/__init__.pyc
   /usr/lib64/python2.7/site-packages/hawkey/test/__init__.pyo
```

in the specfile the `%files` section looks like this:

```
%files
/usr/lib64/python2.7/site-packages/hawkey/__init__.py
%exclude /usr/lib64/python2.7/site-packages/hawkey/__pycache__/
/usr/lib64/python2.7/site-packages/hawkey/_hawkeymodule.so
/usr/lib64/python2.7/site-packages/hawkey/test/__init__.py
%exclude /usr/lib64/python2.7/site-packages/hawkey/test/__pycache__/
/usr/lib64/python2.7/site-packages/hawkey/test/_hawkey_testmodule.so
%exclude /usr/local/include/hawkey/__pycache__/
/usr/local/include/hawkey/advisory.h
/usr/local/include/hawkey/advisorypkg.h
/usr/local/include/hawkey/advisoryref.h
/usr/local/include/hawkey/errno.h
/usr/local/include/hawkey/goal.h
/usr/local/include/hawkey/nevra.h
/usr/local/include/hawkey/package.h
/usr/local/include/hawkey/packagelist.h
/usr/local/include/hawkey/packageset.h
/usr/local/include/hawkey/query.h
/usr/local/include/hawkey/reldep.h
/usr/local/include/hawkey/repo.h
/usr/local/include/hawkey/sack.h
/usr/local/include/hawkey/selector.h
/usr/local/include/hawkey/stringarray.h
/usr/local/include/hawkey/subject.h
/usr/local/include/hawkey/types.h
/usr/local/include/hawkey/util.h
/usr/local/include/hawkey/version.h
%exclude /usr/local/lib64/__pycache__/
/usr/local/lib64/libhawkey.so
/usr/local/lib64/libhawkey.so.2
%exclude /usr/local/lib64/pkgconfig/__pycache__/
/usr/local/lib64/pkgconfig/hawkey.pc
%exclude /usr/local/share/man/man3/__pycache__/
/usr/local/share/man/man3/hawkey.3
```
I don't know how we want to approach this, but in terms of debugging and testing some function that would put `%define _unpackaged_files_terminate_build 0` macro to the top of generated specfile would be handy, as it should allow build to continue despite such errors.

**Comment by LukasSlouka** [comment]
Also [this may be usefull](http://jcg.wtf/blog/2010/09/an-annoying-and-non-obvious-rpmbuild-feature-enhancement/)

File: tests/mock_build/test_build_hawkey.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        self.base.build_project()
+        self.base.run_compiled_source_analysis()
+        self.base.install_project()
+        self.base.run_installed_source_analysis()
+        self.base.spec.BuildRequires = self.base.spec.BuildRequires.union(
+            set(['expat-devel', 'rpm-devel',
+                 'zlib-devel', 'check-devel',
+                 'gcc', 'libsolv-tools',
+                 'python-devel', 'python-nose',
+                 'python-sphinx']))
```
Comment (jsilhan) [inline tests/mock_build/test_build_hawkey.py]: `build_rpm_recover` doesn't find python-sphinx and python-nose?

**Comment by jsilhan** [comment]
1) it should be fixed in the upstream by now. If not open an issue (and fix it), please.
2) this should be fixed in python plugin (add .pyo and .pyc files into `%files) https://github.com/rh-lab-q/rpg/issues/253

File: tests/mock_build/test_build_hawkey.py (hunk 1 above)
Comment (cibo94) [inline tests/mock_build/test_build_hawkey.py]: I think he should (for testing purposes) remove python-nose, python-sphinx, gcc to see if build_rpm_recover will find it.

**Comment by jsilhan** [comment]
c903889 merged

**Comment by LukasSlouka** [comment]
`build_rpm_recover` does not find `python-nose`, but it found `python-sphinx`

**Comment by LukasSlouka** [comment]
@cibo94 says that travis/tito fail is due to too many changes and that we need new version.

**Comment by LukasSlouka** [comment]
I just removed flake error, but mock had another problem, build with flake error passed as u can see in build history of rpg on travis. [link to successful build](https://travis-ci.org/rh-lab-q/rpg/builds/76102360)

**Comment by LukasSlouka** [comment]
Finally I found the problem I had. It is in this 
```python
                subprocess.Popen(
                    [
                        "mock", "--no-clean",
                        "--no-cleanup-after",
                        "--verbose",
                        "--root", distro + '-' + arch,
                        "--rebuild", str(srpm),
                        "--resultdir=" + str(self.temp_dir)
                    ],
                    stdout=subprocess.PIPE,
                    stderr=subprocess.STDOUT
                )
```

concretely in `"--root", distro + '-' + arch,` option. It causes rpmbuild to not find files. If I run the test without this option, it fails due to `installed but unpackaged files error` on `.pyc` and `.pyo` files, which I mentioned earlier. After I have added those manually, the rpmbuild passed and created nice and beautifull rpm

**Comment by LukasSlouka** [comment]
This can be merged, test code wont change anymore.

**Comment by jsilhan** [comment]
Great you have found the root issue. Merged, thanks

**Pull request merged**